### PR TITLE
[WIP] Style engine: register styles for enqueuing and rendering

### PR DESCRIPTION
**Pull request by ramonjd**
WIP

## What?

Background context: https://github.com/WordPress/gutenberg/pull/40987#issuecomment-1139281230

## Why?

One of the aims is to be able to bundle output styles, and also open up the possibility of pre-processing, e.g., deduping

## How?

Creating separate methods to handle style registration, taking into account the various "levels" of styles e.g., global and block supports.

## Testing Instructions

Using Empty Theme:

<details>

<summary>Example</summary>

```html
<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"var:preset|color|pale-pink"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"var:preset|color|luminous-vivid-orange"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"#adc63b"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"#7a4857"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"var:preset|color|light-green-cyan"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"#8262b8"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"var:preset|color|luminous-vivid-amber"}}}}} /-->

<!-- wp:site-title {"style":{"elements":{"link":{"color":{"text":"#40b876"}}}}} /-->

```

</details>

Instead of a `<style />` tag for each rule, we're collating them into one:

```html
	<style>
.wp-elements-6d26dce62e274fbec11380003187995e a { color: var(--wp--preset--color--pale-pink); }
.wp-elements-a6211bd0f50d510936cd495c9e6faf09 a { color: var(--wp--preset--color--luminous-vivid-orange); }
.wp-elements-22d4684938a5ddc0b4a8acf9de8f342c a { color: #adc63b; }
.wp-elements-ba2415e4f4b296c98a0588d746879f5e a { color: #7a4857; }
.wp-elements-ec2ade59a6ecee7bacbe0d1db8c47e65 a { color: var(--wp--preset--color--light-green-cyan); }
.wp-elements-f5f55867de2f30d0db781fb9ec6877b6 a { color: #8262b8; }
.wp-elements-413f8142f63e0ef2e72d3770f1ef248a a { color: var(--wp--preset--color--luminous-vivid-amber); }
.wp-elements-75e03a47eff33fd77f6c021bd12f9034 a { color: #40b876; }
</style>
```

The plan is to allow opt-in support for cascade layers as well:

```html
<style id='wp-styles-layers-inline-css'>
@layer block-supports;
@layer block-supports {
	.wp-elements-bcb267d41f505e6ed08295db66596d26 a {
		color: #b5a6c2;
	}
	.wp-elements-51918bcc051f54979165325dde6fb588 a {
		color: var(--wp--preset--color--vivid-red);
	}
	.wp-elements-18f8c54e1a82f75d4d71f249a8dd88f4 a {
		color: var(--wp--preset--color--luminous-vivid-amber);
	}
}
</style>
```

### CODE REVIEW

**Comment by aristath** [comment]
This looks very promising! 👍

File: packages/style-engine/class-wp-style-engine.php [hunk 1]
```diff
@@ -513,3 +552,23 @@ function wp_style_engine_generate( $block_styles, $options = array() ) {
...
+ *
+ * @param array $block_styles An array of styles from a block's attributes.
+ * @param array $options An array of options to determine the output.
+ * @return void
+ */
+function wp_style_engine_enqueue_block_support_styles( $block_styles, $options = array() ) {
+	if ( class_exists( 'WP_Style_Engine' ) ) {
+		$style_engine = WP_Style_Engine::get_instance();
+		// Get style rules, then register/enqueue.
+		$new_options = array_merge( $options, array( 'enqueue' => true ) );
```
Comment (ramonjd) [inline packages/style-engine/class-wp-style-engine.php]: Try [wp_parse_args](https://developer.wordpress.org/reference/functions/wp_parse_args/) here

File: packages/style-engine/class-wp-style-engine-renderer.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+		}
+		$css_rule_block .= $options['prettify'] ? "$indent}\n" : ' }';
+		return $css_rule_block;
+	}
+
+	// @TODO Using cascade layers should be opt-in.
+	/**
+	 * Builds layers and styles rules from registered layers and styles for output.
+	 */
+	public static function enqueue_cascade_layers() {
```
Comment (ramonjd) [inline packages/style-engine/class-wp-style-engine-renderer.php:126]: Just testing outputting cascade layers for now to test the concept. 

We really want registered styles to be the focus beforehand.

TODO:
- Needs to be tested against other styles for specificity
- How to make this opt-in? (We're instantiating `WP_Style_Engine` as a singleton, so should it be passed to the constructor somehow)

File: packages/style-engine/class-wp-style-engine.php [hunk 3]
```diff
@@ -513,3 +566,27 @@ function wp_style_engine_generate( $block_styles, $options = array() ) {
...
+/**
+ * Global public interface to register block support styles support to be output in the frontend.
+ *
+ * @access public
+ *
+ * @param array $block_styles An array of styles from a block's attributes.
+ * @param array $options An array of options to determine the output.
+ * @return void
+ */
+function wp_style_engine_enqueue_styles( $block_styles, $options = array() ) {
```
Comment (ramonjd) [inline packages/style-engine/class-wp-style-engine.php:580]: Here we're generating and enqueuing block styles, a.k.a values from a styles object `{ spacing: { ... }, typography: { ... }, ... }` 

So maybe it should be called `wp_style_engine_enqueue_block_styles`.

File: packages/style-engine/class-wp-style-engine-renderer.php [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+	 * @return string A filtered CSS property. Empty if not allowed.
+	 */
+	public static function sanitize_property_declaration( $property_declaration ) {
+		return esc_html( safecss_filter_attr( $property_declaration ) );
+	}
+
+	/**
+	 * Creates a string consisting of CSS property declarations suitable for the value of an HTML element's style attribute.
+	 *
+	 * @param array $css_definitions An collection of CSS definitions `[ [ 'color' => 'red' ] ]`.
```
Comment (ramonjd) [inline packages/style-engine/class-wp-style-engine-renderer.php:62]: References to `$css_definitions` should rather be `$css_declarations` since they contain `property: value` pairs

File: lib/block-supports/elements.php [hunk 5]
```diff
@@ -105,17 +105,13 @@ function gutenberg_render_elements_support_styles( $pre_render, $block ) {
 	$link_block_styles = isset( $element_block_styles['link'] ) ? $element_block_styles['link'] : null;
 
 	if ( $link_block_styles ) {
-		$styles = gutenberg_style_engine_generate(
+		gutenberg_style_engine_enqueue_styles(
```
Comment (aristath) [inline lib/block-supports/elements.php:108]: The reason this draft PR is hard to follow is because this function is missing and not defined anywhere 😅

File: packages/style-engine/class-wp-style-engine.php (hunk 3 above)
Comment (aristath) [reply packages/style-engine/class-wp-style-engine.php:580]: That would make sense, yes 👍

File: lib/block-supports/elements.php (hunk 5 above)
Comment (ramonjd) [reply lib/block-supports/elements.php:108]: Ha. You got me. 

This stuff should be better documented. I'll address that in the README.md file. 

So it's _actually_ calling a renamed `wp_style_engine_enqueue_styles`, found [here](https://github.com/WordPress/gutenberg/pull/41424/files#diff-69616e5e4ff1a06362db0b7165c330b00614d4ea0d3a28c80003c0860ab8b93fR580).

The `gutenberg_` prefix is added during the plugin build step via some [webpack voodoo](https://github.com/WordPress/gutenberg/blob/trunk/tools/webpack/packages.js#L61). 

So internally, the plugin fires `gutenberg_style_engine_enqueue_styles` while in WordPress core it'll eventually fire `wp_style_engine_enqueue_styles`. It's a migration pattern that's used throughout the plugin.
